### PR TITLE
Deep flavour PR

### PR DESCRIPTION
**Pull request by mverzett**
This PR introduces the DeepCSV and DeepCMVA (disabled) taggers into the RECO sequences. 

   * The tagger outputs are not stored in the default PAT sequence yet as they would make it crash due to the lack of the TagInfo in the AOD event. All basic tests on my side are passing.
   * This PR depends on [external data](https://github.com/cms-data/RecoBTag-Combined/pull/4) being available in the IB.
   * @slava77 suggested memory and CPU footprint checks, I was hoping jenkins would take care of it.
   * We would like to back-port this PR (disabling the changes to the RECO format) once this is approved. Can I already make a backport PR?

### CODE REVIEW

**Comment by slava77** [comment]
@mverzett jenkins does not provide memory checks
Based on the size of the .json file  I'm not expecting a large increase in memory

**Comment by slava77** [comment]
I'll wait for the next IB to start the tests. The .json files should show up in the -2300 IB

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
looks like I started the tests too early: jenkins was using -1100 IB which didn't have the training files

**Comment by mverzett** [comment]
@slava77 can you poke CMS bot again? I don't think I have the authority to do it myself

@cmsbuild please test

**Comment by monttj** [comment]
@cmsbuild please test

**Comment by ferencek** [comment]
@mverzett, while newly introduced files look ok, in the modified files the indentation for newly added lines look messed up. Why is that?

**Comment by mverzett** [comment]
@ferencek can you point me to the problems? Is it c++ or python? Python should have given problem in that case.

**Comment by ferencek** [comment]
@mverzett, have you given up on defining the ShallowTagInfo in such a way it would allow removing the BoostedDoubleSVTagInfo and replacing it with the ShallowTagInfo?

**Comment by ferencek** [comment]
Both C++ and Python. Just look at the file diffs in https://github.com/cms-sw/cmssw/pull/17247/files

**Comment by mverzett** [comment]
@ferencek I believed is due to the longstanding debate between space and tab indentation. No matter how I set emacs to behave, some files will be OK some will not. Any suggestion of how to set emacs to comply to whichever standard is used in the current file is more than welcome, of course

**Comment by ferencek** [comment]
@mverzett, I've abandoned emacs long time ago and I don't have much experience with it to suggest a specific setting. What is strange to me is that it appears that emacs displays tabs as spaces to you and you are then not even aware that the code indentation is messed up. Can you configure emacs to display tabs as 8 spaces according to the usual convention?

For the ShallowTagInfo and BoostedDoubleSVTagInfo I agree that they can be merged later in a separate PR.

**Comment by slava77** [comment]
the unit test error is in
```
      [c] Input file root://eoscms.cern.ch//eos/cms/store/user/cmsbuild/store/relval/CMSSW_8_0_0/RelValTTbar_13/MINIAODSIM/PU25ns_80X_mcRun2_asymptotic_v4-v1/10000/A65CD249-BFDA-E511-813A-0025905A6066.root could not be opened.
```

I suggest to ignore this failing unit test for now

**Comment by mverzett** [comment]
@slava77 I think I should somehow take care of @ferencek comments, but any suggestion on how is welcome. Also, I would like to first gather your comments as well, so I change the code only once, reducing the chances of messing up.

**Comment by slava77** [comment]
On 1/25/17 5:46 AM, Mauro Verzetti wrote:
> @slava77 <https://github.com/slava77> I think I should somehow take care
> of @ferencek <https://github.com/ferencek> comments, but any suggestion
> on how is welcome. Also, I would like to first gather your comments as
> well, so I change the code only once, reducing the chances of messing up.
> 


I didn't have time yet to look at the code.

Please proceed with fixes to address Dinko's comments first.

Thank you.

**Comment by mverzett** [comment]
@ferencek , the last commit should fix the indentation

**Comment by ferencek** [comment]
@mverzett, thanks for fixing the indentation. In `DataFormats/BTauReco/interface/TaggingVariable.h` and `DataFormats/BTauReco/src/classes_def.xml ` I still see some misalignment wrt the rest of the of the code but that is less critical.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by mverzett** [comment]
@slava77 any updates on this PR?

**Comment by slava77** [comment]
@mverzett 
I have just started my review.
Sorry for the delay.

File: DataFormats/BTauReco/src/TaggingVariable.cc [hunk 1]
```diff
@@ -141,8 +144,11 @@ const char* const TaggingVariableTokens[] = {
   /* [jetPt]                                    = */ "jetPt",
```
Comment (slava77) [inline DataFormats/BTauReco/src/TaggingVariable.cc]: I don't find descriptions for the new Jet_ variables

File: DataFormats/BTauReco/src/classes_def.xml [hunk 2]
```diff
@@ -71,6 +71,18 @@
   <class name="reco::BoostedDoubleSVTagInfoRefVector"/>
   <class name="edm::Wrapper<reco::BoostedDoubleSVTagInfoCollection>"/>
 
+  <class name="reco::ShallowTagInfo" ClassVersion="4">
+		<version ClassVersion="3" checksum="3123030258"/>
+		<version ClassVersion="4" checksum="3904302444"/>
```
Comment (slava77) [inline DataFormats/BTauReco/src/classes_def.xml]: please remove the intermediate versions. Change the latest version to 3
```xml
 <class name="reco::ShallowTagInfo" ClassVersion="3">
	<version ClassVersion="3" checksum="3904302444"/>
```

File: PhysicsTools/PatAlgos/plugins/PATJetProducer.cc [hunk 3]
```diff
@@ -100,6 +100,11 @@ PATJetProducer::PATJetProducer(const edm::ParameterSet& iConfig)  :
         if ((pos !=  std::string::npos) && (pos != label.length() - 7)) {
             label.erase(pos+7); // trim a tail after "JetTags"
         }
+        if(it->instance().size()) {
+            std::stringstream name;
+            name << label << ":" << it->instance();
+            label = name.str();
```
Comment (slava77) [inline PhysicsTools/PatAlgos/plugins/PATJetProducer.cc]: these are all strings
wouldn't the following work ```label+string(":")+it->instance()```
?

File: PhysicsTools/PatAlgos/python/producersLayer1/jetProducer_cfi.py [hunk 4]
```diff
@@ -48,7 +48,19 @@
         cms.InputTag("pfCombinedMVAV2BJetTags"),
         # CTagging
         cms.InputTag('pfCombinedCvsLJetTags'),
-        cms.InputTag('pfCombinedCvsBJetTags')
+        cms.InputTag('pfCombinedCvsBJetTags'),
+        # DeepFlavour
+        # cms.InputTag('pfDeepCSVJetTags:probb'),
```
Comment (slava77) [inline PhysicsTools/PatAlgos/python/producersLayer1/jetProducer_cfi.py]: is this commented out code needed?
Please remove or add comments inline in the code why the commented out block is relevant

[I can guess that we need to wait for a reference release inputs to be able to pass the PAT unit tests]

File: PhysicsTools/PatAlgos/python/tools/jetTools.py [hunk 5]
```diff
@@ -279,6 +279,30 @@ def setupBTagging(process, jetSource, pfCandidates, explicitJTA, pvSource, svSou
...
+                if svClustering or fatJets != cms.InputTag(''):
+                    setupSVClustering(getattr(process, btagPrefix+btagInfo+labelName+postfix), svClustering, algo, rParam, fatJets, groomedFatJets)
+            if btagInfo == 'pfDeepCSVPositiveTagInfos':
+                setattr(process, btagPrefix+btagInfo+labelName+postfix, btag.pfDeepCSVPositiveTagInfos.clone(svTagInfos = cms.InputTag(btagPrefix+'pfInclusiveSecondaryVertexFinderTagInfos'+labelName+postfix)))
+                if svClustering or fatJets != cms.InputTag(''):
+                    setupSVClustering(getattr(process, btagPrefix+btagInfo+labelName+postfix), svClustering, algo, rParam, fatJets, groomedFatJets)
+            if btagInfo == 'pfDeepCMVATagInfos':
+                setattr(process, btagPrefix+btagInfo+labelName+postfix, btag.pfDeepCMVATagInfos.clone(pfDeepCSVTagInfos = cms.InputTag(btagPrefix+'pfDeepCSVTagInfos'+labelName+postfix), ipInfoSrc = cms.InputTag(btagPrefix+"pfImpactParameterTagInfos"+labelName+postfix), muInfoSrc = cms.InputTag(btagPrefix+"softPFMuonsTagInfos"+labelName+postfix), elInfoSrc = cms.InputTag(btagPrefix+"softPFElectronsTagInfos"+labelName+postfix)))
```
Comment (slava77) [inline PhysicsTools/PatAlgos/python/tools/jetTools.py]: some line breaks here and a few other places in this file would be nice; maybe in a separate PR

File: RecoBTag/Combined/plugins/DeepCMVATagInfoProducer.cc [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+#include "DataFormats/BTauReco/interface/CandIPTagInfo.h"
+#include "DataFormats/BTauReco/interface/CandSecondaryVertexTagInfo.h"
+#include "DataFormats/BTauReco/interface/ShallowTagInfo.h"
+#include "DataFormats/BTauReco/interface/TaggingVariable.h"
+#include "RecoBTag/SecondaryVertex/interface/CombinedSVComputer.h"
+#include "RecoBTau/JetTagComputer/interface/JetTagComputer.h"
+#include "RecoBTau/JetTagComputer/interface/JetTagComputerRecord.h"
+#include "DataFormats/BTauReco/interface/JetTag.h"
+
+#include <math.h>
```
Comment (slava77) [inline RecoBTag/Combined/plugins/DeepCMVATagInfoProducer.cc]: ```<cmath>``` please

File: RecoBTag/Combined/plugins/DeepCMVATagInfoProducer.cc [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+DeepCMVATagInfoProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup)
+{	
+	// get input TagInfos from DeepCSV
+	edm::Handle< std::vector<reco::ShallowTagInfo> > nnInfos;
+	iEvent.getByToken(deepNNSrc_, nnInfos);
+	
+	// get other Taginfos
+	edm::Handle< edm::View<BaseTagInfo> > ipInfos;
+	iEvent.getByToken(ipInfoSrc_, ipInfos);
+	std::vector<const BaseTagInfo*> ipBaseInfo;
```
Comment (slava77) [inline RecoBTag/Combined/plugins/DeepCMVATagInfoProducer.cc]: this looks spurious

File: RecoBTag/Combined/python/DeepFlavourJetTagsProducer_cfi.py [hunk 8]
```diff
@@ -0,0 +1,13 @@
+import FWCore.ParameterSet.Config as cms
+
+pfDeepCSVJetTags = cms.EDProducer(
```
Comment (slava77) [inline RecoBTag/Combined/python/DeepFlavourJetTagsProducer_cfi.py]: preferably, keep one producer per cfi file
and keep the same name for the file base as the producer name (e.g. pfDeepCSVJetTags_cfi and pfDeepCMVAJetTags_cfi)

File: RecoBTag/Combined/python/DeepNNTagInfoProducer_cfi.py [hunk 9]
```diff
@@ -0,0 +1,8 @@
+import FWCore.ParameterSet.Config as cms
+from RecoBTag.SecondaryVertex.combinedSecondaryVertexCommon_cff import combinedSecondaryVertexCommon
+
+pfDeepCSVTagInfos = cms.EDProducer(
```
Comment (slava77) [inline RecoBTag/Combined/python/DeepNNTagInfoProducer_cfi.py]: match the file base to pfDeepCSVTagInfos

**Comment by slava77** [comment]
@mverzett 
I'm done with the visual inspection
the comments are mostly cosmetic.
Please address the classVersion and the header include; and the rest at will
Thanks.

**Comment by mverzett** [comment]
@slava77, thanks for letting me know. I will go through your comments and fix everything or ask clarifications

**Comment by slava77** [comment]
Here are some plots from a data test in 2016E JetHT, showing some meaningful values of the discriminants, just taking udsg and bb as extremes

prob udsg
all_origvssign837_runjetht2016ewf136p761c_recojetedmreftobaseprodtofloatsassociationvector_pfdeepcsvjettags_probudsg_rereco_obj_data_171

prob bb
all_origvssign837_runjetht2016ewf136p761c_recojetedmreftobaseprodtofloatsassociationvector_pfdeepcsvjettags_probc_rereco_obj_atdata__size

On the technical side, I do not observe any issues: 
- CPU time is small 
- size per event is about 500 bytes on 200 events in 5 discriminants
- there is no significant increase in memory

So, what remains is the follow up to the code review.

File: DataFormats/BTauReco/src/TaggingVariable.cc (hunk 1 above)
Comment (mverzett) [inline DataFormats/BTauReco/src/TaggingVariable.cc]: @slava77 What do you mean? The descriptions are in the right handside in the strings.

**Comment by mverzett** [comment]
@slava77 I'm in the process of applying your comments. Given the IB I was working on is gone, how do you suggest to move the PR code to avoid spurious merge commits in the git history? Thanks

**Comment by ferencek** [comment]
@mverzett, why not rebase all the commits to the latest IB?

**Comment by mverzett** [comment]
@ferencek The last time I tried rebasing I made a mess, can you make me a working example?

**Comment by boudoul** [comment]
hi @mverzett , look at this nice step by step recipe from Slava : 
 https://github.com/cms-sw/cmssw/issues/11426

**Comment by mverzett** [comment]
@boudoul , thanks for the reference!

File: DataFormats/BTauReco/src/TaggingVariable.cc (hunk 1 above)
Comment (slava77) [inline DataFormats/BTauReco/src/TaggingVariable.cc]: The comment is for the TaggingVariableDescription array above. There is no description for , e.g. 
Jet_SoftMu which should be right below z_ratio on L135

**Comment by slava77** [comment]
On 2/3/17 4:37 PM, Mauro Verzetti wrote:
> @slava77 <https://github.com/slava77> I'm in the process of applying
> your comments. Given the IB I was working on is gone, how do you suggest
> to move the PR code to avoid spurious merge commits in the git history?
> Thanks

you will need to rebase
https://twiki.cern.ch/twiki/bin/viewauth/CMS/SWGuideHATS2016Git#CMSSW_rebasing_recipe

**Comment by mverzett** [comment]
Hi, thanks for the instruction. When I try to do it I get:
```
>>> git fetch official-cmssw
error: refs/remotes/official-cmssw/test-new-prs does not point to a valid object!
fatal: Unable to find remote helper for 'https'
```
I guess it's normal, but how do I solve it? Thanks

**Comment by slava77** [comment]
well, you can do the rebase in the new IB as well

```
git cms-addpkg for all relevant packages
git fetch my-cmssw DeepFlavourPR-from-CMSSW_9_0_X:DeepFlavourPR-from-CMSSW_9_0_X
git checkout DeepFlavourPR-from-CMSSW_9_0_X
# the rest should be just as in the instructions
```

**Comment by slava77** [comment]
.. at least I'm guessing it should work

your commit history looks flat. So, you should be able to replay the commits by cherry-picking

``` 
git fetch my-cmssw DeepFlavourPR-from-CMSSW_9_0_X:DeepFlavourPR-from-CMSSW_9_0_X
git cherry-pick `git merge-base DeepFlavourPR-from-CMSSW_9_0_X CMSSW_9_0_X_2017-IBNUMBERHERE`..DeepFlavourPR-from-CMSSW_9_0_X 
```

**Comment by mverzett** [comment]
@slava77 , indeed cheery-picking was my back-up solution. I will try it and update everything. Thanks

**Comment by slava77** [comment]
@mverzett 
please clarify if the latest update also addresses the code review comments and that you are done so far with updates.
Thanks.

**Comment by mverzett** [comment]
@slava77 Sorry I did not do it before.
This last set of commits addresses both your comments and the ones from @ferencek and should be complete and working. The only comment I did not address are the better description of the tagging variables, as some of them were not introduced in this PR and would require some digging to find out.

File: DataFormats/BTauReco/src/TaggingVariable.cc (hunk 1 above)
Comment (slava77) [inline DataFormats/BTauReco/src/TaggingVariable.cc]: May I suggest to enter at least some dummy values in TaggingVariableDescription
https://github.com/cms-sw/cmssw/pull/17247/files#diff-89cfcec978581c291c1e7822bb06e964R137

I don't care as much about the content of the description. The problem is that there are no values set,
which can lead to crashes if there is some code using descriptions for printouts or something.

Comment (mverzett) [inline DataFormats/BTauReco/src/TaggingVariable.cc]: @slava77 I know see what you mean. I fixed it.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/17247	6f1e3b6
- code changes are in line with the PR description and the follow up code review; new b-tagging producers/products are added
- jenkins tests pass and comparisons with baseline show no difference (the new products are not monitored)
- local tests were done with the earlier version of this PR and show roughly expected performance in the saved quantities and rather small impact on CPU and memory https://github.com/cms-sw/cmssw/pull/17247#issuecomment-276869452

**Comment by mverzett** [comment]
@slava77 , can I start a backport to 80X? I got a bit lost for the re-miniAOD schedule, am I still in time? (guess not, but worth trying)

**Comment by slava77** [comment]
On 2/6/17 10:36 AM, Mauro Verzetti wrote:
> @slava77 <https://github.com/slava77> , can I start a backport to 80X? I
> got a bit lost for the re-miniAOD schedule, am I still in time? (guess
> not, but worth trying)

It would be more practical to wait until this PR is merged in 90X to
submit a backport,
just in case you receive more comments.

re-mniAOD  started last Friday.

Since this PR changes data and MC, it will wait for the data+MC re-miniAOD.
I'm expecting that it will happen in connection with the legacy data rereco.
The target for this is some time in March.

**Comment by slava77** [comment]
@monttj 
please check this PR, your signature is needed here.
Thank you.

**Comment by monttj** [comment]
+1
@mverzett thank you for new development!

**Comment by davidlange6** [comment]
+1

**Pull request merged**